### PR TITLE
Add `Action.getenv` and use variadic options in `New()`.

### LINKED ISSUE #18: Expose a field in `Action` with `getenvFunc`.
I am happy to do the work here, I just wanted sign off before sending a PR. I'm also happy to conform to your preferences, some possible design choices:

- Make `Action.GetEnv` an exported field **OR** make `Action.getEnv` unexported
- Support an options pattern in `New()` so that `GetEnv` can be optionally set **OR** provide a `NewWithGetEnv()` function similar to `NewWithWriter()` **OR** provide a `Action.WithGetEnv()` method for setting the field
- Check if `Action.GetEnv` is `nil` before usage and fallback to `os.Getenv` (would require a wrapper method) **OR** trust that `Action.GetEnv: os.Getenv` will be the default set in `New()` and leave it at that

----

**Motivation**: This makes it somewhat easier to write unit tests for GitHub actions.

Just now I am testing an "input parsing" function and ended up defining

```go
type getInputer interface {
	GetInput(string) string
}

func NewFromInputs(action getInputer) (*Config, error) {
	// ...
}
```

so that in my tests I can use

```go
type GetInputMap struct {
	Map map[string]string
}

func (gim GetInputMap) GetInput(i string) string {
	return gim.Map[i]
}
```

### PR DESCRIPTION
**Pull request by dhermes**
Fixes #18.

### CODE REVIEW

File: actions.go [hunk 1]
```diff
@@ -88,15 +93,11 @@ func (c *Action) IssueCommand(cmd *Command) {
 // with the 'Command' argument as it's scope is unclear in the current
 // TypeScript implementation.
 func (c *Action) IssueFileCommand(cmd *Command) error {
-	return c.issueFileCommand(cmd, os.Getenv)
```
Comment (dhermes) [inline actions.go:91]: I did a bunch of editorializing (i.e. removing the internal implementations that take a `getenvFunc`) but the changes can be removed if you want. I isolated the changes in small commits.

File: actions_test.go [hunk 2]
```diff
@@ -79,7 +95,7 @@ func TestAction_AddMask(t *testing.T) {
 	t.Parallel()
 
 	var b bytes.Buffer
-	a := NewWithWriter(&b)
+	a := New(OptWriter(&b))
```
Comment (dhermes) [inline actions_test.go:98]: Ditto here on editorializing; I did this to prepare for just straight up deleting `NewWithWriter()`. I figured it'd make sense to leave it in for at least one commit to `main` to give people a chance to migrate to `New(opts...)`.

File: actions.go [hunk 3]
```diff
@@ -53,22 +53,27 @@ const (
...
+	}
+	return a
 }
 
 // NewWithWriter creates a wrapper using the given writer. This is useful for
 // tests. The given writer cannot add any prefixes to the string, since GitHub
 // requires these special strings to match a very particular format.
 func NewWithWriter(w io.Writer) *Action {
-	return &Action{w: w}
+	return New(OptWriter(w))
```
Comment (dhermes) [inline actions.go:68]: Would you also want to accept `opts ...Option` to the args in `NewWithWriter()` or just let it languish in preparation for removal?

File: actions.go [hunk 4]
```diff
@@ -53,22 +53,27 @@ const (
 
 // New creates a new wrapper with helpers for outputting information in GitHub
 // actions format.
-func New() *Action {
-	return &Action{w: os.Stdout}
+func New(opts ...Option) *Action {
+	a := &Action{w: os.Stdout, getenv: os.Getenv}
+	for _, opt := range opts {
+		opt(a)
```
Comment (sethvargo) [inline actions.go:59]: Can we add a guard that `opt != nil` before calling? Sometimes developers do weird things...

File: actions.go (hunk 3 above)
Comment (sethvargo) [reply actions.go:68]: I like what you've done here. Can you add:

```go
//
// Deprecated: Use New() with WithWriter instead.
```

File: options.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+// has no external dependencies and provides a Go-like interface for interacting
+// with GitHub Actions' build system.
+package githubactions
+
+import (
+	"io"
+)
+
+// Option is a modifier for an Action
+type Option func(*Action)
```
Comment (sethvargo) [inline options.go:25]: I think options should take the action _and return the action_:

```go
type Option func(*Action) *Action
```

Gives us more flex in the future.

File: options.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+// OptWriter sets the writer function on an Action. By default, this will
+// be `os.Stdout` from the standard library.
+func OptWriter(w io.Writer) Option {
```
Comment (sethvargo) [inline options.go:29]: Nit: I prefer `With*` naming for options:

- `WithWriter`
- `WithFields`
- `WithGetEnv`

File: options.go [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+// OptFields sets the extra command field on an Action.
+func OptFields(fields CommandProperties) Option {
+	return func(a *Action) {
+		a.fields = fields
+	}
+}
+
+// OptGetenv sets the `Getenv` function on an Action. By default, this will
+// be `os.Getenv` from the standard library.
+func OptGetenv(getenv getenvFunc) Option {
```
Comment (sethvargo) [inline options.go:44]: Since `getenvFunc` is now part of a public API, I think we also need to make it public.

**Review by sethvargo — CHANGES_REQUESTED** [review]

**Review by ashutoshgngwr — APPROVED** [review]
**My two cents:** Implementation is really neat. The code cleanup is also appreciable. I'd have named the `Option` functions with `With` prefix, e.g. WithWrite, but that's just me. :sweat_smile:

File: actions.go (hunk 1 above)
Comment (sethvargo) [reply actions.go:91]: This is fine.

File: actions_test.go (hunk 2 above)
Comment (sethvargo) [reply actions_test.go:98]: Also fine

File: actions.go (hunk 4 above)
Comment (dhermes) [reply actions.go]: Sure thing. Done.

File: actions.go (hunk 3 above)
Comment (dhermes) [reply actions.go:73]: Done.

File: options.go (hunk 6 above)
Comment (dhermes) [reply options.go:29]: Sure thing, will rename.

Comment (dhermes) [reply options.go]: Done.

File: options.go (hunk 7 above)
Comment (dhermes) [reply options.go]: SGTM

Comment (dhermes) [reply options.go]: Done.

File: options.go (hunk 5 above)
Comment (dhermes) [reply options.go]: Done

**Review by sethvargo — APPROVED** [review]

**Pull request merged**